### PR TITLE
Use CacheableTaskQueue for state generation and caching

### PR DESCRIPTION
**Pull request by ajsutton**
## PR Description
Follow on PR from #2628 that moves state generation over to use `CacheableTaskQueue` replacing `StateGenerationQueue`.  Removes duplicate code and closes a small gap between checking the state cache and starting regeneration which could result in recreating a state that had just finished being recreated.

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

### CODE REVIEW

**Review by rolfyone — APPROVED** [review]
LGTM

File: ethereum/core/src/main/java/tech/pegasys/teku/core/stategenerator/StateGenerationTask.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+  @Override
+  public Stream<Bytes32> streamIntermediateSteps() {
+    final Bytes32 rootParent = tree.getParent(tree.getRootHash()).orElseThrow();
+    return Stream.iterate(
+            tree.getParent(blockRoot),
+            Optional::isPresent,
+            current -> current.flatMap(tree::getParent))
+        .flatMap(Optional::stream)
+        // The root's parent is in the tree but is worse that the current starting point so exclude
+        .filter(ancestorRoot -> !ancestorRoot.equals(rootParent));
```
Comment (mbaxter) [inline ethereum/core/src/main/java/tech/pegasys/teku/core/stategenerator/StateGenerationTask.java:71]: To make it clearer what's happening here, maybe we should add some validation in the constructor:
```
checkArgument(tree.getRootHash().equals(baseBlockAndState.getRoot()), "Tree must be rooted at the base block");
```

File: storage/src/main/java/tech/pegasys/teku/storage/store/StoreBuilder.java [hunk 2]
```diff
@@ -106,16 +107,17 @@ public static StoreBuilder forkChoiceStoreBuilder(
   }
 
   private void createDefaults() {
-    if (stateGenerationQueue == null) {
+    if (stateTaskQueue == null) {
       checkState(stateAndBlockProvider != null, "StateAndBlockProvider must be defined");
```
Comment (mbaxter) [inline storage/src/main/java/tech/pegasys/teku/storage/store/StoreBuilder.java]: Now that we're passing `stateAndBlockProvider` into `Store` directly, I think we need to add this assert to `assertValid()` as well.

File: storage/src/main/java/tech/pegasys/teku/storage/store/StoreBuilder.java [hunk 3]
```diff
@@ -106,16 +107,17 @@ public static StoreBuilder forkChoiceStoreBuilder(
...
-      stateGenerationQueue = StateGenerationQueue.create(stateAndBlockProvider, metricsSystem);
+      stateTaskQueue =
+          CachingTaskQueue.create(metricsSystem, "memory_states", storeConfig.getStateCacheSize());
```
Comment (mbaxter) [inline storage/src/main/java/tech/pegasys/teku/storage/store/StoreBuilder.java]: We could probably just construct this inside of `Store`.

**Review by mbaxter — COMMENTED** [review]
LGTM!

File: storage/src/main/java/tech/pegasys/teku/storage/store/StoreBuilder.java (hunk 2 above)
Comment (ajsutton) [reply storage/src/main/java/tech/pegasys/teku/storage/store/StoreBuilder.java]: Done.

File: storage/src/main/java/tech/pegasys/teku/storage/store/StoreBuilder.java (hunk 3 above)
Comment (ajsutton) [reply storage/src/main/java/tech/pegasys/teku/storage/store/StoreBuilder.java]: Yep, makes sense - then it follows the same pattern as for checkpoint states.

File: ethereum/core/src/main/java/tech/pegasys/teku/core/stategenerator/StateGenerationTask.java (hunk 1 above)
Comment (ajsutton) [reply ethereum/core/src/main/java/tech/pegasys/teku/core/stategenerator/StateGenerationTask.java:71]: Done.

**Pull request merged**